crypto/pkcs7/pkcs7.c

```diff
@@ -608,3 +609,490 @@ void PKCS7_RECIP_INFO_get0_alg(PKCS7_RECIP_INFO *ri, X509_ALGOR **penc) {
+      content = p7->d.signed_and_enveloped->enc_data->enc_data;
+      if (content == NULL) {
+        content = ASN1_OCTET_STRING_new();
+        if (content == NULL) {
+          OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+          goto err;
+        }
+        p7->d.signed_and_enveloped->enc_data->enc_data = content;
+      }
+      break;
+    case NID_pkcs7_enveloped:
+      content = p7->d.enveloped->enc_data->enc_data;
+      if (content == NULL) {
+        content = ASN1_OCTET_STRING_new();
+        if (content == NULL) {
+          OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+          goto err;
+        }
+        p7->d.enveloped->enc_data->enc_data = content;
+      }
+      break;
+    case NID_pkcs7_signed:
+      si_sk = p7->d.sign->signer_info;
+      OPENSSL_BEGIN_ALLOW_DEPRECATED
+      content = PKCS7_get_octet_string(p7->d.sign->contents);
+      OPENSSL_END_ALLOW_DEPRECATED
+      // If detached data then the content is excluded
+      OPENSSL_BEGIN_ALLOW_DEPRECATED
+      if (PKCS7_type_is_data(p7->d.sign->contents) && PKCS7_is_detached(p7)) {
+        OPENSSL_END_ALLOW_DEPRECATED
+        ASN1_OCTET_STRING_free(content);
+        content = NULL;
+        p7->d.sign->contents->d.data = NULL;
+      }
+      break;
+
+    case NID_pkcs7_digest:
+      content = PKCS7_get_octet_string(p7->d.digest->contents);
+      // If detached data, then the content is excluded
+      OPENSSL_BEGIN_ALLOW_DEPRECATED
+      if (PKCS7_type_is_data(p7->d.digest->contents) && PKCS7_is_detached(p7)) {
+        OPENSSL_END_ALLOW_DEPRECATED
+        ASN1_OCTET_STRING_free(content);
+        content = NULL;
+        p7->d.digest->contents->d.data = NULL;
+      }
+      break;
+
+    default:
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CONTENT_TYPE);
+      goto err;
+  }
+
+  if (si_sk != NULL) {
+    for (size_t ii = 0; ii < sk_PKCS7_SIGNER_INFO_num(si_sk); ii++) {
+      si = sk_PKCS7_SIGNER_INFO_value(si_sk, ii);
+      if (si->pkey == NULL) {
+        continue;
+      }
+      int sign_nid = OBJ_obj2nid(si->digest_alg->algorithm);
+      bio_tmp = PKCS7_find_digest(&md_ctx, bio_tmp, sign_nid);
+      if (bio_tmp == NULL) {
+        goto err;
+      }
+      // We now have the EVP_MD_CTX, lets do the signing.
+      if (!EVP_MD_CTX_copy_ex(md_ctx_tmp, md_ctx)) {
+        goto err;
+      }
+      // We don't currently support signed attributes
+      if (sk_X509_ATTRIBUTE_num(si->auth_attr) > 0) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_PKCS7_DATASIGN);
+        goto err;
+      }
+      unsigned char *abuf = NULL;
+      unsigned int abuflen = EVP_PKEY_size(si->pkey);
+      if (abuflen == 0 || (abuf = OPENSSL_malloc(abuflen)) == NULL) {
+        goto err;
+      }
+      if (!EVP_SignFinal(md_ctx_tmp, abuf, &abuflen, si->pkey)) {
+        OPENSSL_free(abuf);
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_EVP_LIB);
+        goto err;
+      }
+      ASN1_STRING_set0(si->enc_digest, abuf, abuflen);
```

Comment:
There's apparently no test coverage of the latter part of this loop.

Reply:
Correct. that will come once we add PKCS7 verification and expand PKCS7 signing. that will happen 2 PRs from now. In the meantime, I have a "parent" PR [here](https://github.com/aws/aws-lc/pull/1993) that combines all changes across the PR series (pre-cleanup for each individual PR, so larger and messier). The coverage report [shows](https://app.codecov.io/gh/aws/aws-lc/pull/1993?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=aws#0ee70fa588380281c0522614f1db15ca-R1024) it's covered there, but we should drop a TODO here to cover (I generally avoid TODO's across commits, but there are currently none in the `crypto/pkcs7/` directory, so it won't get lost in any noise).